[Form] Fixed check of violation constraint #12792

xelaris opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #12792 |
| License | MIT |
| Doc PR |  |

ConstraintViolation::getConstraint() must not expect to provide a constraint as long as Symfony\Component\Validator\ExecutionContext exists (before 3.0)

Tobion commented:
Please provide a test case

xelaris commented:
tests added

Tobion commented:
ping @webmozart

stof commented on src/Symfony/Component/Form/Tests/Extension/Validator/EventListener/ValidationListenerTest.php:
```diff
@@ -66,7 +66,7 @@ protected function setUp()
     private function getConstraintViolation($code = null)
     {
-        return new ConstraintViolation($this->message, $this->messageTemplate, $this->params, null, 'prop.path', null, null, $code, new Form());
+        return $this->getMock('Symfony\Component\Validator\ConstraintViolation', array('getConstraint'), array($this->message, $this->messageTemplate, $this->params, null, 'prop.path', null, null, $code));
```
no need to use a mock IMO. Just add an argument in the getter to decide whether the last argument should be the constraint or ``null``

xelaris replied:
this was my first step. Than I realized that it is probably a good idea to test, if `getConstraint` is called by `validateForm`. Otherwise it could be simply removed without breaking the tests (which would also fix the bug). So I changed it to build a mock and added `$violation->expects($this->atLeastOnce())` to the test cases.

Tobion replied:
if `getConstraint` is not called, then other tests should fail. don't mix both things.

xelaris committed:
[Form] Fixed check of violation constraint #12792

ConstraintViolation::getConstraint() must not expect to provide a
constraint as long as Symfony\Component\Validator\ExecutionContext
exists (before 3.0)

xelaris commented:
As @Tobion said, I have changed the commit so that it is discrete and only test the particular case.

fabpot commented:
Thank you @xelaris.
